src/Generating/EnderDragonFightStructuresGen.cpp

```diff
@@ -0,0 +1,282 @@
+
+#include "Globals.h"
+
+#include "EnderDragonFightStructuresGen.h"
+#include "../Chunk.h"
+#include "../Entities/EnderCrystal.h"
+#include "../Mobs/EnderDragon.h"
+#include "../WorldStorage/SchematicFileSerializer.h"
+
+const std::array<Vector3i, 48> cEnderDragonFightStructuresGen::m_CagePos =
+{
+	{
+		// First layer
+		{-2, -1, -2}, {-2, -1, -1}, {-2, -1, 0}, {-2, -1, 1}, {-2, -1, 2},
+		{2,  -1, -2}, {2,  -1, -1}, {2,  -1, 0}, {2,  -1, 1}, {2,  -1, 2},
+		{-1, -1, -2}, {0,  -1, -2}, {1,  -1, -2},
+		{-1, -1, 2},  {0,  -1, 2},  {1,  -1, 2},
+
+		// Second layer
+		{-2, 0, -2}, {-2, 0, -1}, {-2, 0, 0}, {-2, 0, 1}, {-2, 0, 2},
+		{2,  0, -2}, {2,  0, -1}, {2,  0, 0}, {2,  0, 1}, {2,  0, 2},
+		{-1, 0, -2}, {0,  0, -2}, {1,  0, -2},
+		{-1, 0, 2},  {0,  0, 2},  {1,  0, 2},
+
+		// Third layer
+		{-2, 1, -2}, {-2, 1, -1}, {-2, 1, 0}, {-2, 1, 1}, {-2, 1, 2},
+		{2,  1, -2}, {2,  1, -1}, {2,  1, 0}, {2,  1, 1}, {2,  1, 2},
+		{-1, 1, -2}, {0,  1, -2}, {1,  1, -2},
+		{-1, 1, 2},  {0,  1, 2},  {1,  1, 2},
+	}
+};
+
+
+
+
+
+const std::array<Vector3i, 26> cEnderDragonFightStructuresGen::m_CageAir =
+{
+	{
+		// First layer
+		{-1, -1, -1}, {0, -1, -1}, {1, -1, -1},
+		{-1, -1, 1},  {0, -1, 1},  {1, -1, 1},
+		{-1, -1, 0}, {1, -1, 0},
+
+		// Second layer
+		{-1, 0, -1}, {0, 0, -1}, {1, 0, -1},
+		{-1, 0, 1},  {0, 0, 1},  {1, 0, 1},
+		{-1, 0, 0}, {1, 0, 0}, {0, 0, 0},
+
+		// Third layer
+		{-1, 1, -1}, {0, 1, -1}, {1, 1, -1},
+		{-1, 1, 1},  {0, 1, 1},  {1, 1, 1},
+		{-1, 1, 0}, {1, 1, 0}, {0, 1, 0},
+	}
+};
+
+
+
+
+
+cEnderDragonFightStructuresGen::cEnderDragonFightStructuresGen(int a_Seed) :
+		m_Noise(a_Seed)
+{
+}
+
+
+
+
+
+void cEnderDragonFightStructuresGen::Init(const AString & a_TowerProperties, int a_Radius)
+{
+	// Loads the fountain schematic
+	if (!cSchematicFileSerializer::LoadFromSchematicFile(
+		m_Fountain, AString("Prefabs") + cFile::GetPathSeparator() + "SinglePieceStructures" + cFile::GetPathSeparator() + "EndFountain.schematic"))
+	{
+		LOGWARNING("EnderDragonFightStructuresGen is missing it's end fountain prefab, please update your cuberite server files! There will be no end fountain!");
+	}
+	// Reads the given tower properties
+	const auto TowerPropertiesVector = StringSplitAndTrim(a_TowerProperties, ";");
+	std::vector<sTowerProperties> TowerProperties;
+	for (const auto & TowerProperty : TowerPropertiesVector)
+	{
+		const auto TowerPropertyVector = StringSplitAndTrim(TowerProperty, "|");
+		if (TowerPropertyVector.size() != 3)
+		{
+			LOGWARNING("Got unknown parameters on generating obsidian pillars: %s, Please use \"Height|Radius|HasCage\"; ...", TowerProperty);
+			continue;
+		}
+		int Height = std::min(std::stoi(TowerPropertyVector[0]), cChunkDef::Height - 2);  // The highest block placed is two blocks above the given height (the cage above some towers)
```

Comment:
Try saying const auto MaxHeight = cChunkDef::Height - 2, then using MaxHeight. Seems like cChunkDef::Height is being treated as a true constant literal, no symbol emitted to the linker (yay!) but std::min takes arguments by reference (and literals have no address).

Reply:
Weird though, considering std::min asks for _const_ references, and clearly you can pass normal integral literals to it without complaint...

Reply:
Huh, apparently it's just the language.

```cpp
struct S {
    static const int x = 0; // static data member
    // a definition outside of class is required if it is odr-used
};
const int& f(const int& r);
 
int n = b ? (1, S::x) // S::x is not odr-used here
          : f(S::x);  // S::x is odr-used here: a definition is required

              ^^^^ 😮
```

https://en.cppreference.com/w/cpp/language/definition#ODR-use